src/core/thread/network_data_notifier.cpp

```diff
@@ -152,6 +169,75 @@ void Notifier::HandleCoapResponse(Error aResult)
     }
 }
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+
+bool Notifier::IsEligibleForRouterRoleUpgrade(void) const
+{
+    bool    isEligible = false;
+    uint8_t activeRouterCount;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsChild() && Get<Mle::MleRouter>().IsRouterEligible());
+
+    // RouterUpgradeThreshold can be explicitly set to zero in some of
+    // cert tests to disallow device to become router.
+
+    VerifyOrExit(Get<Mle::MleRouter>().GetRouterUpgradeThreshold() != 0);
+
+    // Check the local network data to ensure that we are a border
+    // router providing IP connectivity and therefore eligible to
+    // request router role upgrade with `kBorderRouterRequst`
+    // status.
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
+
+    activeRouterCount = Get<RouterTable>().GetActiveRouterCount();
+    VerifyOrExit((activeRouterCount >= Get<Mle::MleRouter>().GetRouterUpgradeThreshold()) &&
+                 (activeRouterCount < Mle::kMaxRouters));
+
+    VerifyOrExit(Get<Leader>().CountBorderRouters(kRouterRoleOnly) < Mle::kRouterUpgradeBorderRouterRequestThreshold);
+    isEligible = true;
+
+exit:
+    return isEligible;
+}
+
+void Notifier::RequestRouterRoleIfEligible(void)
```

Comment:
```suggestion
void Notifier::ScheduleRequestRouterRoleIfEligible(void)
```
since we only schedule it